### PR TITLE
Ziti tunnel.cfgtype.v2.0

### PR DESCRIPTION
**Pull request by scareything**
Add service configuration types to the database that will support VoIP and SCTP use cases.

The _intercept-v1_ config type will be used for tunneled services that:
- intercept CIDRs and/or multiple addresses
- intercept multiple ports and/or port ranges
- use identity dial
- require source IP spoofing at the hosting tunneler

Some example configuration documents:

passive FTP with control port=21 and data ports={44000-45000}:
```json
{
    "protocols": [ "tcp" ],
    "addresses": [ "ftp.pasv.forme" ],
    "portRanges": [
        { "low": 21, "high": 21 },
        { "low": 44000, "high": 45000 }
    ]
}
```

SIP
```json
{
    "protocols": [ "tcp", "udp" ],
    "addresses": [ "sip-server.ecorp", "10.1.0.0/24", "10.1.2.0/24" ],
    "portRanges": [ { "low": 5060, "high": 5061 } ],
    "dialOptions": {
        "identity": "$intercepted_ip.tag[ipv4Address]"
    },
    "sourceIp": "$tunneler_id.tag[ipv4Address]"
}
```

RTP
```json
{
    "protocols": [ "tcp", "udp" ],
    "addresses": [ "10.1.0.0/24", "10.1.2.0/24" ],
    "portRanges": [ { "low": 6970, "high": 6999 } ],
    "dialOptions": {
        "identity": "$intercepted_ip"
    },
    "listenOptions": {
        "identity": "$tunneler_id.tag[ipv4Address]"
    },
    "sourceIp": "$tunneler_id.tag[ipv4Address]"
}
```

<details><summary>Here's the raw intercept-v1 schema</summary>
<p>

```json
{
    "$id": "http://ziti-edge.netfoundry.io/schemas/intercept-v1.schema.json",
    "$defs": {
        "ipAddressFormat": {
            "oneOf": [
                { "format": "ipv4" },
                { "format": "ipv6" }
            ]
        },
        "ipAddress": {
            "type": "string",
            "$ref": "#/$defs/ipAddressFormat"
        },
        "cidr": {
            "type": "string",
            "oneOf": [
                { "pattern": "^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(\\/(3[0-2]|[1-2][0-9]|[0-9]))$" },
                { "pattern": "^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(\\/(12[0-8]|1[0-1][0-9]|[1-9][0-9]|[0-9]))$" }
            ]
        },
        "hostname": {
            "type": "string",
            "format": "hostname",
            "not": { "$ref": "#/$defs/ipAddressFormat" }
        },
        "address": {
            "oneOf": [
                { "$ref": "#/$defs/ipAddress" },
                { "$ref": "#/$defs/hostname" },
                { "$ref": "#/$defs/cidr" }
            ]
        },
        "portNumber": {
            "type": "integer",
            "minimum": 0,
            "maximum": 65535
        },
        "portRange": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "low":  { "$ref": "#/$defs/portNumber" },
                "high": { "$ref": "#/$defs/portNumber" }
            }
        },
        "protocolName": {
            "type": "string",
            "enum": [ "tcp", "udp", "sctp" ]
        },
        "timeoutSeconds": {
            "type": "integer",
            "minimum": 0,
            "maximum": 2147483647
        },
        "inhabitedSet": {
            "type": "array",
            "minItems": 1,
            "uniqueItems": true
        }
    },
    "type": "object",
    "properties": {
        "protocols": {
            "allOf": [
                { "$ref": "#/$defs/inhabitedSet" },
                { "items": { "$ref": "#/$defs/protocolName" } }
            ]
        },
        "addresses": {
            "allOf": [
                { "$ref": "#/$defs/inhabitedSet" },
                { "items": { "$ref": "#/$defs/address" } }
            ]
        },
        "portRanges": {
            "allOf": [
                { "$ref": "#/$defs/inhabitedSet" },
                { "items": { "$ref": "#/$defs/portRange" } }
            ]
        },
        "dialOptions": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "identity": {
                    "type": "string",
                    "description": "Dial a terminator with the specified identity. '$intercepted_protocol', '$intercepted_ip', '$intercepted_port are resolved to the corresponding value of the intercepted address."
                },
                "connectTimeoutSeconds": {
                    "$ref": "#/$defs/timeoutSeconds",
                    "description": "defaults to 5"
                }
            }
        },
        "sourceIp": {
            "type": "string",
            "description": "The source IP to spoof when the connection is egressed from the hosting tunneler. '$tunneler_id.name' resolves to the name of the client tunneler's identity. '$tunneler_id.tag[tagName]' resolves to the value of the 'tagName' tag on the client tunneler's identity."
        },
        "destinationPort" {
            "description": "The port that the hosting tunneler will dial to complete the connection."
        }
    },
    "additionalProperties": false,
    "required": [
        "protocols",
        "addresses",
        "portRanges"
    ]
}
```
</p>
</details>

The new configuration for hosting services mainly adds bind options and the ability to dial whatever address was intercepted at the client tunneler in addition to dialing a predetermined address.

Here's the familiar "predetermined address" scenario:

```json
{
    "protocol": "tcp",
    "address": "1.2.3.4",
    "port": 1234
}

```

And here's a _host_ configuration that dials the IP address and port that was intercepted at the client:

```json
{
    "dialInterceptedProtocol": true,
    "dialInterceptedAddress": true,
    "dialInterceptedPort": true
}
```

So if the client was intercepting 10.0.0.0/16 on ports 7000-7010 and a client connected to 10.0.1.12:7002/tcp, the host configuration above would dial the server at 10.0.1.12:7002.

<details><summary>Finally, here's the raw host-v1 schema that the go map was derived from:</summary>
<p>

```json
{
    "$id": "http://ziti-edge.netfoundry.io/schemas/host-v1.schema.json",
    "$defs": {
        "ipAddressFormat": {
            "oneOf": [
                { "format": "ipv4" },
                { "format": "ipv6" }
            ]
        },
        "ipAddress": {
            "type": "string",
            "$ref": "#/$defs/ipAddressFormat"
        },
        "hostname": {
            "type": "string",
            "format": "hostname",
            "not": { "$ref": "#/$defs/ipAddressFormat" }
        },
        "address": {
            "oneOf": [
                { "$ref": "#/$defs/ipAddress" },
                { "$ref": "#/$defs/hostname" }
            ]
        }
    },
    "type": "object",
    "properties": {
        "protocol": {
            "type": "string",
            "enum": [ "tcp", "udp", "sctp" ],
            "description": "Dial the specified protocol when a ziti client connects to the service."
        },
        "dialInterceptedProtocol": {
            "type": "boolean",
            "enum": [ true ],
            "description": "Dial the same protocol that was intercepted at the client tunneler. 'protocol' and 'dialInterceptedProtocol' are mutually exclusive."
        },
        "address": {
            "$ref": "#/$defs/address",
            "description": "Dial the specified ip address or hostname when a ziti client connects to the service."
        },
        "dialInterceptedAddress": {
            "type": "boolean",
            "enum": [ true ],
            "description": "Dial the same ip address that was intercepted at the client tunneler. 'address' and 'dialInterceptedAddress' are mutually exclusive."
        },
        "port": {
            "type": "integer",
            "minimum": 0,
            "maximum": 65535,
            "description": "Dial the specified port when a ziti client connects to the service."
        },
        "dialInterceptedPort": {
            "type": "boolean",
            "enum": [ true ],
            "description": "Dial the same port that was intercepted at the client tunneler. 'port' and 'dialInterceptedPort' are mutually exclusive."
        },
        "listenOptions": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "cost": {
                    "type": "integer",
                    "minimum": 0,
                    "maximum": 65535,
                    "description": "defaults to 0"
                },
                "precedence": {
                    "type": "string",
                    "enum": [ "default", "required", "failed" ],
                    "description": "defaults to 'default'"
                },
                "connectTimeoutSeconds": {
                    "type": "integer",
                    "minimum": 0,
                    "maximum": 2147483647,
                    "description": "defaults to 5"
                },
                "maxConnections": {
                    "type": "integer",
                    "minimum": 1,
                    "description": "defaults to 3"
                },
                "identity": {
                    "type": "string",
                    "description": "Associate the hosting terminator with the specified identity. '$tunneler_id.name' resolves to the name of the hosting tunneler's identity. '$tunneler_id.tag[tagName]' resolves to the value of the 'tagName' tag on the hosting tunneler's identity."
                },
                "bindUsingEdgeIdentity": {
                    "type": "boolean",
                    "description": "Associate the hosting terminator with the name of the hosting tunneler's identity. Setting this to 'true' is equivalent to setting 'identiy=$tunneler_id.name'"
                }
            }
        }
    },
    "additionalProperties": false,
    "allOf": [
        {
            "oneOf": [
                { "required": [ "protocol" ] },
                { "required": [ "dialInterceptedProtocol" ] }
            ]
        },
        {
            "oneOf": [
                { "required": [ "address" ] },
                { "required": [ "dialInterceptedAddress" ] }
            ]
        },
        {
            "oneOf": [
                { "required": [ "port" ] },
                { "required": [ "dialInterceptedPort" ] }
            ]
        }
    ]
}
```
</p>
</details>

### CODE REVIEW

**Comment by ekoby** [comment]
not sure I like `tunneler_id` in the listen options (and others), why not just `$id`?
are other fields supported -- `$id.name`, or `$id.id`?

**Comment by scareything** [comment]
> not sure I like `tunneler_id` in the listen options (and others), why not just `$id`?
> are other fields supported (other than `tags`) -- `$id.name`, or `$id.id`?

I don't feel strongly about `tunneler_id`, but the goal was to make it clear that we are talking about the identity that the tunneler is using. It didn't feel like `id` was quite descriptive enough in this context, but again I don't feel strongly.

As for fields, for now I thought `$foo.name` and `$foo.tag[tagName]` would be more than we need. I was thinking that $foo without a field would be an error, but I guess it could be a synonym for $foo.name.

File: controller/persistence/migration_v13.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	"math"
+)
+
+func (m *Migrations) createTunnelerClientV2dot0ConfigType(step *boltz.MigrationStep) {
+	configName := "ziti-tunneler-client.v2.0"
+	clientConfigV2dot0TypeId := "g7cIWbcGg"
+	clientConfigTypeV2dot0 := &ConfigType{
+		BaseExtEntity: boltz.BaseExtEntity{Id: clientConfigV2dot0TypeId},
+		Name:          configName,
+		Schema: map[string]interface{}{
```
Comment (andrewpmartinez) [inline controller/persistence/migration_v13.go:15]: Might have been easier to write this in JSON and then parse it to a map[string]interface{}.

Comment (scareything) [reply controller/persistence/migration_v13.go:15]: This is what I did. The raw schema is in this PRs description. I was wondering if there was a good place to check in the raw schema file also?

**Review by andrewpmartinez — APPROVED** [review]

File: controller/persistence/migration_v13.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+					"description": "The source IP to spoof when the connection is egressed from the hosting tunneler. '$tunneler_id.name' resolves to the name of the client tunneler's identity. '$tunneler_id.tag[tagName]' resolves to the value of the 'tagName' tag on the client tunneler's identity.",
+				},
+			},
+			"required": []interface{}{
+				"protocols",
+				"addresses",
+				"portRanges",
+			},
+		},
+	}
```
Comment (andrewpmartinez) [inline controller/persistence/migration_v13.go:162]: The schema looks sane, but it is quite a bit.

**Review by smilindave26 — APPROVED** [review]

**Pull request merged**